Flaky test: `TestMetadataFileUpdatedAgentRestart`

### Summary
`TestMetadataFileUpdatedAgentRestart` sometimes fails asynchronously after completion.

### Description
A goroutine causes a failure after the test completes, sometimes.  It looks like there's an unmocked call to `StateManager.ForceSave`

### Expected Behavior
Tests pass

### Observed Behavior
Tests sometimes fail, inconsistently

### Environment Details
TravisCI

### Supporting Log Snippets
```
==================
WARNING: DATA RACE

Read at 0x00c420246044 by goroutine 605:
  testing.(*common).Fail()
      /home/travis/.gimme/versions/go1.7.linux.amd64/src/testing/testing.go:411 +0x9f
  testing.(*common).FailNow()
      /home/travis/.gimme/versions/go1.7.linux.amd64/src/testing/testing.go:431 +0x38
  testing.(*common).Fatalf()
      /home/travis/.gimme/versions/go1.7.linux.amd64/src/testing/testing.go:496 +0x94
  testing.(*T).Fatalf()
      <autogenerated>:12 +0x86
  github.com/aws/amazon-ecs-agent/agent/vendor/github.com/golang/mock/gomock.(*Controller).Call()
      /home/travis/gopath/src/github.com/aws/amazon-ecs-agent/agent/vendor/github.com/golang/mock/gomock/controller.go:132 +0x733
  github.com/aws/amazon-ecs-agent/agent/statemanager/mocks.(*MockStateManager).ForceSave()
      /home/travis/gopath/src/github.com/aws/amazon-ecs-agent/agent/statemanager/mocks/statemanager_mocks.go:45 +0xa8
  github.com/aws/amazon-ecs-agent/agent/engine.(*DockerTaskEngine).createContainer()
      github.com/aws/amazon-ecs-agent/agent/engine/_test/_obj_test/docker_task_engine.go:780 +0xf85
  github.com/aws/amazon-ecs-agent/agent/engine.(*DockerTaskEngine).(github.com/aws/amazon-ecs-agent/agent/engine.createContainer)-fm()
      github.com/aws/amazon-ecs-agent/agent/engine/_test/_obj_test/docker_task_engine.go:142 +0x9a
  github.com/aws/amazon-ecs-agent/agent/engine.(*DockerTaskEngine).applyContainerState()
      github.com/aws/amazon-ecs-agent/agent/engine/_test/_obj_test/docker_task_engine.go:1073 +0x35a
  github.com/aws/amazon-ecs-agent/agent/engine.(*DockerTaskEngine).transitionContainer()
      github.com/aws/amazon-ecs-agent/agent/engine/_test/_obj_test/docker_task_engine.go:1043 +0x90
  github.com/aws/amazon-ecs-agent/agent/engine.(*managedTask).progressContainers.func1()
      github.com/aws/amazon-ecs-agent/agent/engine/_test/_obj_test/task_manager.go:487 +0xba

Previous write at 0x00c420246044 by goroutine 688:
  testing.tRunner.func1()
      /home/travis/.gimme/versions/go1.7.linux.amd64/src/testing/testing.go:605 +0x34a
  testing.tRunner()
      /home/travis/.gimme/versions/go1.7.linux.amd64/src/testing/testing.go:612 +0xe5

Goroutine 605 (running) created at:
  github.com/aws/amazon-ecs-agent/agent/engine.(*managedTask).startContainerTransitions()
      github.com/aws/amazon-ecs-agent/agent/engine/_test/_obj_test/task_manager.go:536 +0x20e
  github.com/aws/amazon-ecs-agent/agent/engine.(*managedTask).progressContainers()
      github.com/aws/amazon-ecs-agent/agent/engine/_test/_obj_test/task_manager.go:490 +0x2f6
  github.com/aws/amazon-ecs-agent/agent/engine.(*managedTask).overseeTask()
      github.com/aws/amazon-ecs-agent/agent/engine/_test/_obj_test/task_manager.go:132 +0x945

Goroutine 688 (finished) created at:
  testing.(*T).Run()
      /home/travis/.gimme/versions/go1.7.linux.amd64/src/testing/testing.go:646 +0x52f
  testing.RunTests.func1()
      /home/travis/.gimme/versions/go1.7.linux.amd64/src/testing/testing.go:793 +0xb9
  testing.tRunner()
      /home/travis/.gimme/versions/go1.7.linux.amd64/src/testing/testing.go:610 +0xc9
  testing.RunTests()
      /home/travis/.gimme/versions/go1.7.linux.amd64/src/testing/testing.go:799 +0x4ba
  testing.(*M).Run()
      /home/travis/.gimme/versions/go1.7.linux.amd64/src/testing/testing.go:743 +0x12f
  main.main()
      github.com/aws/amazon-ecs-agent/agent/engine/_test/_testmain.go:348 +0x2f8

==================

panic: Fail in goroutine after TestMetadataFileUpdatedAgentRestart has completed
goroutine 11945 [running]:
panic(0xac1d00, 0xc420437390)
	/home/travis/.gimme/versions/go1.7.linux.amd64/src/runtime/panic.go:500 +0x1ae
testing.(*common).Fail(0xc420246000)
	/home/travis/.gimme/versions/go1.7.linux.amd64/src/testing/testing.go:412 +0x182
testing.(*common).FailNow(0xc420246000)
	/home/travis/.gimme/versions/go1.7.linux.amd64/src/testing/testing.go:431 +0x39
testing.(*common).Fatalf(0xc420246000, 0xb98f7e, 0x2e, 0xc4205f6b90, 0x5, 0x5)
	/home/travis/.gimme/versions/go1.7.linux.amd64/src/testing/testing.go:496 +0x95
testing.(*T).Fatalf(0xc420246000, 0xb98f7e, 0x2e, 0xc4205f6b90, 0x5, 0x5)
	<autogenerated>:12 +0x87
github.com/aws/amazon-ecs-agent/agent/vendor/github.com/golang/mock/gomock.(*Controller).Call(0xc4205740e0, 0xb10360, 0xc420456920, 0xb8499e, 0x9, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/home/travis/gopath/src/github.com/aws/amazon-ecs-agent/agent/vendor/github.com/golang/mock/gomock/controller.go:132 +0x734
github.com/aws/amazon-ecs-agent/agent/statemanager/mocks.(*MockStateManager).ForceSave(0xc420456920, 0x35, 0xc420212330)
	/home/travis/gopath/src/github.com/aws/amazon-ecs-agent/agent/statemanager/mocks/statemanager_mocks.go:45 +0xa9
github.com/aws/amazon-ecs-agent/agent/engine.(*DockerTaskEngine).createContainer(0xc4200c10e0, 0xc420634000, 0xc4202ff520, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	github.com/aws/amazon-ecs-agent/agent/engine/_test/_obj_test/docker_task_engine.go:780 +0xf86
github.com/aws/amazon-ecs-agent/agent/engine.(*DockerTaskEngine).(github.com/aws/amazon-ecs-agent/agent/engine.createContainer)-fm(0xc420634000, 0xc4202ff520, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	github.com/aws/amazon-ecs-agent/agent/engine/_test/_obj_test/docker_task_engine.go:142 +0x9b
github.com/aws/amazon-ecs-agent/agent/engine.(*DockerTaskEngine).applyContainerState(0xc4200c10e0, 0xc420634000, 0xc4202ff520, 0x2, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	github.com/aws/amazon-ecs-agent/agent/engine/_test/_obj_test/docker_task_engine.go:1073 +0x35b
github.com/aws/amazon-ecs-agent/agent/engine.(*DockerTaskEngine).transitionContainer(0xc4200c10e0, 0xc420634000, 0xc4202ff520, 0xc400000002)
	github.com/aws/amazon-ecs-agent/agent/engine/_test/_obj_test/docker_task_engine.go:1043 +0x91
github.com/aws/amazon-ecs-agent/agent/engine.(*managedTask).progressContainers.func1(0xc4202ff520, 0xc400000002)
	github.com/aws/amazon-ecs-agent/agent/engine/_test/_obj_test/task_manager.go:487 +0xbb
created by github.com/aws/amazon-ecs-agent/agent/engine.(*managedTask).startContainerTransitions
	github.com/aws/amazon-ecs-agent/agent/engine/_test/_obj_test/task_manager.go:536 +0x20f
```

Here's another one from https://github.com/aws/amazon-ecs-agent/pull/906:

```
panic: Fail in goroutine after TestMetadataFileUpdatedAgentRestart has completed

goroutine 11993 [running]:
panic(0xad1600, 0xc420312ad0)
/home/travis/.gimme/versions/go1.7.linux.amd64/src/runtime/panic.go:500 +0x1ae
testing.(*common).Fail(0xc4201f9c80)
/home/travis/.gimme/versions/go1.7.linux.amd64/src/testing/testing.go:412 +0x182
testing.(*common).FailNow(0xc4201f9c80)
/home/travis/.gimme/versions/go1.7.linux.amd64/src/testing/testing.go:431 +0x39
testing.(*common).Fatalf(0xc4201f9c80, 0xba9b3e, 0x2e, 0xc4206114a0, 0x5, 0x5)
/home/travis/.gimme/versions/go1.7.linux.amd64/src/testing/testing.go:496 +0x95
testing.(*T).Fatalf(0xc4201f9c80, 0xba9b3e, 0x2e, 0xc4206114a0, 0x5, 0x5)
<autogenerated>:12 +0x87
github.com/aws/amazon-ecs-agent/agent/vendor/github.com/golang/mock/gomock.(*Controller).Call(0xc420652860, 0xb203a0, 0xc4204386f0, 0xb9531b, 0x9, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
/home/travis/gopath/src/github.com/aws/amazon-ecs-agent/agent/vendor/github.com/golang/mock/gomock/controller.go:132 +0x734
github.com/aws/amazon-ecs-agent/agent/statemanager/mocks.(*MockStateManager).ForceSave(0xc4204386f0, 0x35, 0xc420215170)
/home/travis/gopath/src/github.com/aws/amazon-ecs-agent/agent/statemanager/mocks/statemanager_mocks.go:45 +0xa9
github.com/aws/amazon-ecs-agent/agent/engine.(*DockerTaskEngine).createContainer(0xc4202f39e0, 0xc420129600, 0xc4202c7040, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
github.com/aws/amazon-ecs-agent/agent/engine/_test/_obj_test/docker_task_engine.go:798 +0xf86
github.com/aws/amazon-ecs-agent/agent/engine.(*DockerTaskEngine).(github.com/aws/amazon-ecs-agent/agent/engine.createContainer)-fm(0xc420129600, 0xc4202c7040, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
github.com/aws/amazon-ecs-agent/agent/engine/_test/_obj_test/docker_task_engine.go:142 +0x9b
github.com/aws/amazon-ecs-agent/agent/engine.(*DockerTaskEngine).applyContainerState(0xc4202f39e0, 0xc420129600, 0xc4202c7040, 0xc400000002, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
github.com/aws/amazon-ecs-agent/agent/engine/_test/_obj_test/docker_task_engine.go:1091 +0x35b
github.com/aws/amazon-ecs-agent/agent/engine.(*DockerTaskEngine).transitionContainer(0xc4202f39e0, 0xc420129600, 0xc4202c7040, 0xc400000002)
github.com/aws/amazon-ecs-agent/agent/engine/_test/_obj_test/docker_task_engine.go:1061 +0x91
github.com/aws/amazon-ecs-agent/agent/engine.(*managedTask).progressContainers.func1(0xc4202c7040, 0xc400000002)
github.com/aws/amazon-ecs-agent/agent/engine/_test/_obj_test/task_manager.go:498 +0xbb
created by github.com/aws/amazon-ecs-agent/agent/engine.(*managedTask).startContainerTransitions
github.com/aws/amazon-ecs-agent/agent/engine/_test/_obj_test/task_manager.go:583 +0x41c
FAIL github.com/aws/amazon-ecs-agent/agent/engine 12.473s

Another, from https://github.com/aws/amazon-ecs-agent/pull/1018:

```
panic: Fail in goroutine after TestMetadataFileUpdatedAgentRestart has completed

goroutine 11919 [running]:
panic(0xac1d00, 0xc4205cfe10)
/home/travis/.gimme/versions/go1.7.linux.amd64/src/runtime/panic.go:500 +0x1ae
testing.(*common).Fail(0xc4206d35c0)
/home/travis/.gimme/versions/go1.7.linux.amd64/src/testing/testing.go:412 +0x182
testing.(*common).FailNow(0xc4206d35c0)
/home/travis/.gimme/versions/go1.7.linux.amd64/src/testing/testing.go:431 +0x39
testing.(*common).Fatalf(0xc4206d35c0, 0xb98f7e, 0x2e, 0xc420458410, 0x5, 0x5)
/home/travis/.gimme/versions/go1.7.linux.amd64/src/testing/testing.go:496 +0x95
testing.(*T).Fatalf(0xc4206d35c0, 0xb98f7e, 0x2e, 0xc420458410, 0x5, 0x5)
<autogenerated>:12 +0x87
github.com/aws/amazon-ecs-agent/agent/vendor/github.com/golang/mock/gomock.(*Controller).Call(0xc42063e900, 0xb10360, 0xc420553250, 0xb8499e, 0x9, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
/home/travis/gopath/src/github.com/aws/amazon-ecs-agent/agent/vendor/github.com/golang/mock/gomock/controller.go:132 +0x734
github.com/aws/amazon-ecs-agent/agent/statemanager/mocks.(*MockStateManager).ForceSave(0xc420553250, 0x35, 0xc420258ba0)
/home/travis/gopath/src/github.com/aws/amazon-ecs-agent/agent/statemanager/mocks/statemanager_mocks.go:45 +0xa9
github.com/aws/amazon-ecs-agent/agent/engine.(*DockerTaskEngine).createContainer(0xc420220480, 0xc4201a43c0, 0xc42059c820, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
github.com/aws/amazon-ecs-agent/agent/engine/_test/_obj_test/docker_task_engine.go:780 +0xf86
github.com/aws/amazon-ecs-agent/agent/engine.(*DockerTaskEngine).(github.com/aws/amazon-ecs-agent/agent/engine.createContainer)-fm(0xc4201a43c0, 0xc42059c820, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
github.com/aws/amazon-ecs-agent/agent/engine/_test/_obj_test/docker_task_engine.go:142 +0x9b
github.com/aws/amazon-ecs-agent/agent/engine.(*DockerTaskEngine).applyContainerState(0xc420220480, 0xc4201a43c0, 0xc42059c820, 0xc400000002, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
github.com/aws/amazon-ecs-agent/agent/engine/_test/_obj_test/docker_task_engine.go:1073 +0x35b
github.com/aws/amazon-ecs-agent/agent/engine.(*DockerTaskEngine).transitionContainer(0xc420220480, 0xc4201a43c0, 0xc42059c820, 0xc400000002)
github.com/aws/amazon-ecs-agent/agent/engine/_test/_obj_test/docker_task_engine.go:1043 +0x91
github.com/aws/amazon-ecs-agent/agent/engine.(*managedTask).progressContainers.func1(0xc42059c820, 0xc400000002)
github.com/aws/amazon-ecs-agent/agent/engine/_test/_obj_test/task_manager.go:487 +0xbb
created by github.com/aws/amazon-ecs-agent/agent/engine.(*managedTask).startContainerTransitions
github.com/aws/amazon-ecs-agent/agent/engine/_test/_obj_test/task_manager.go:536 +0x20f
FAIL github.com/aws/amazon-ecs-agent/agent/engine 13.062s

--- a/agent/engine/docker_task_engine_test.go
+++ b/agent/engine/docker_task_engine_test.go
@@ -1537,6 +1537,7 @@ func TestGetTaskByArn(t *testing.T) {
 
 	sleepTask := testdata.LoadTask("sleep5")
 	sleepTaskArn := sleepTask.Arn
+	sleepTask.SetDesiredStatus(api.TaskStopped)
 	taskEngine.AddTask(sleepTask)
 
 	_, found := taskEngine.GetTaskByArn(sleepTaskArn)
@@ -1911,7 +1912,6 @@ func TestPullNormalImage(t *testing.T) {
 // agent starts, container created, metadata file created, agent restarted, container recovered
 // during task engine init, metadata file updated
 func TestMetadataFileUpdatedAgentRestart(t *testing.T) {
-
 	conf := &defaultConfig
 	conf.ContainerMetadataEnabled = true
 
@@ -1925,12 +1925,14 @@ func TestMetadataFileUpdatedAgentRestart(t *testing.T) {
 	taskEngine, _ := privateTaskEngine.(*DockerTaskEngine)
 	assert.True(t, taskEngine.cfg.ContainerMetadataEnabled, "ContainerMetadataEnabled set to false.")
 
+	taskEngine._time = nil
 	taskEngine.SetSaver(saver)
 	state := taskEngine.State()
 	task := testdata.LoadTask("sleep5")
 
 	container, _ := task.ContainerByName("sleep5")
 	assert.False(t, container.MetadataFileUpdated)
+	container.SetKnownStatus(api.ContainerRunning)
 
 	dockerContainer := &api.DockerContainer{DockerID: containerID, Container: container}
 
@@ -1938,17 +1940,17 @@ func TestMetadataFileUpdatedAgentRestart(t *testing.T) {
 	expectedDockerID := dockerContainer.DockerID
 	expectedContainerName := container.Name
 
+	state.AddTask(task)
 	state.AddContainer(dockerContainer, task)
 
 	client.EXPECT().Version()
 	eventStream := make(chan DockerContainerChangeEvent)
 	client.EXPECT().ContainerEvents(gomock.Any()).Return(eventStream, nil)
-	imageManager.EXPECT().AddAllImageStates(gomock.Any()).AnyTimes()
-	imageManager.EXPECT().RecordContainerReference(gomock.Any()).AnyTimes()
-	imageManager.EXPECT().GetImageStateFromImageName(gomock.Any()).AnyTimes()
-	client.EXPECT().PullImage(gomock.Any(), gomock.Any()).AnyTimes()
-	client.EXPECT().DescribeContainer(gomock.Any()).AnyTimes()
+	client.EXPECT().DescribeContainer(gomock.Any())
+	imageManager.EXPECT().RecordContainerReference(gomock.Any())
+
 	saver.EXPECT().Save().AnyTimes()
+	saver.EXPECT().ForceSave().AnyTimes()
 
 	metadataManager.EXPECT().Update(gomock.Any(), gomock.Any(), gomock.Any()).Do(func(dockerID string, taskARN string, containerName string) {
 		assert.Equal(t, expectedTaskARN, taskARN)

